### PR TITLE
Kandinsky2.2

### PR DESCRIPTION
**Pull request by cene555**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [x] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yiyixuxu** [comment]
@cene555 
Thanks so much for the PR!!! 
didn't expect it, so it's like Christmas comes early 😅🎅🎁

are you still working on the PR, or is it ready for review? let us know

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+                noise_pred, _ = noise_pred.split(latents.shape[1], dim=1)
+
+            # compute the previous noisy sample x_t -> x_t-1
+            latents = self.scheduler.step(
+                noise_pred,
+                t,
+                latents,
+                # YiYi notes: only reason this pipeline can't work with unclip scheduler is that can't pass down this argument
+                #             need to use DDPM scheduler instead
+                # prev_timestep=prev_timestep,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder.py:300]:
```suggestion
```
we can remove this note now:)

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+        new_h += 1
+    new_w = w // scale_factor**2
+    if w % scale_factor**2 != 0:
+        new_w += 1
+    return new_h * scale_factor, new_w * scale_factor
+
+
+class Kandinsky2_2_DecoderControlnetPipeline(DiffusionPipeline):
+    """
+    Pipeline for text-to-image generation using Kandinsky
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py:47]:
```suggestion
    Pipeline for text-to-image generation using Kandinsky with ControlNet guidance.
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+    Args:
+        text_encoder ([`MultilingualCLIP`]):
+            Frozen text-encoder.
+        tokenizer ([`XLMRobertaTokenizer`]):
+            Tokenizer of class
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py:54]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+        generator: Optional[Union[torch.Generator, List[torch.Generator]]] = None,
+        latents: Optional[torch.FloatTensor] = None,
+        output_type: Optional[str] = "pil",
+        return_dict: bool = True,
+    ):
+        """
+        Function invoked when calling the pipeline for generation.
+        Args:
+            prompt (`str` or `List[str]`):
+                The prompt or prompts to guide the image generation.
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py:178]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+            hint (`torch.FloatTensor`):
+                The controlnet condition.
+            image_embeds (`torch.FloatTensor` or `List[torch.FloatTensor]`):
+                The clip image embeddings for text prompt, that will be used to condition the image generation.
+            negative_image_embeds (`torch.FloatTensor` or `List[torch.FloatTensor]`):
+                The clip image embeddings for negative text prompt, will be used to condition the image generation.
+            negative_prompt (`str` or `List[str]`, *optional*):
+                The prompt or prompts not to guide the image generation. Ignored when not using guidance (i.e., ignored
+                if `guidance_scale` is less than `1`).
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py:187]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+                noise_pred, _ = noise_pred.split(latents.shape[1], dim=1)
+
+            # compute the previous noisy sample x_t -> x_t-1
+            latents = self.scheduler.step(
+                noise_pred,
+                t,
+                latents,
+                # YiYi notes: only reason this pipeline can't work with unclip scheduler is that can't pass down this argument
+                #             need to use DDPM scheduler instead
+                # prev_timestep=prev_timestep,
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet.py:288]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py [hunk 7]
```diff
@@ -0,0 +1,385 @@
...
+    arr = np.array(pil_image.convert("RGB"))
+    arr = arr.astype(np.float32) / 127.5 - 1
+    arr = np.transpose(arr, [2, 0, 1])
+    image = torch.from_numpy(arr).unsqueeze(0)
+    return image
+
+
+class Kandinsky2_2_DecoderControlnetImg2ImgPipeline(DiffusionPipeline):
+    """
+    Pipeline for text-to-image generation using Kandinsky
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py:58]:
```suggestion
    Pipeline for image-to-image generation using Kandinsky with ControlNet guidance.
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py [hunk 8]
```diff
@@ -0,0 +1,385 @@
...
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+    Args:
+        text_encoder ([`MultilingualCLIP`]):
+            Frozen text-encoder.
+        tokenizer ([`XLMRobertaTokenizer`]):
+            Tokenizer of class
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py:65]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py [hunk 9]
```diff
@@ -0,0 +1,385 @@
...
+                f"Cannot duplicate `image` of batch size {init_latents.shape[0]} to {batch_size} text prompts."
+            )
+        else:
+            init_latents = torch.cat([init_latents], dim=0)
+
+        shape = init_latents.shape
+        noise = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+
+        # get latents
+        try:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py:149]: need to add add_noise function to unclip scheduler

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py [hunk 10]
```diff
@@ -0,0 +1,385 @@
...
+                noise_pred = torch.cat([noise_pred, variance_pred_text], dim=1)
+
+            if not (
+                hasattr(self.scheduler.config, "variance_type")
+                and self.scheduler.config.variance_type in ["learned", "learned_range"]
+            ):
+                noise_pred, _ = noise_pred.split(latents.shape[1], dim=1)
+
+            # compute the previous noisy sample x_t -> x_t-1
+            try:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_controlnet_img2img.py:355]: I don't understand this here: do we have a scheduler.step that does not accept `generator` argument? 

in any case we should be very clear about what schedulers this pipeline can work with and be able to work with them

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py [hunk 11]
```diff
@@ -0,0 +1,382 @@
...
+    arr = np.array(pil_image.convert("RGB"))
+    arr = arr.astype(np.float32) / 127.5 - 1
+    arr = np.transpose(arr, [2, 0, 1])
+    image = torch.from_numpy(arr).unsqueeze(0)
+    return image
+
+
+class Kandinsky2_2_DecoderImg2ImgPipeline(DiffusionPipeline):
+    """
+    Pipeline for text-to-image generation using Kandinsky
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py:58]:
```suggestion
    Pipeline for image-to-image generation using Kandinsky
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py [hunk 12]
```diff
@@ -0,0 +1,382 @@
...
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+    Args:
+        text_encoder ([`MultilingualCLIP`]):
+            Frozen text-encoder.
+        tokenizer ([`XLMRobertaTokenizer`]):
+            Tokenizer of class
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py:65]:
```suggestion
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py [hunk 13]
```diff
@@ -0,0 +1,382 @@
...
+                f"Cannot duplicate `image` of batch size {init_latents.shape[0]} to {batch_size} text prompts."
+            )
+        else:
+            init_latents = torch.cat([init_latents], dim=0)
+
+        shape = init_latents.shape
+        noise = randn_tensor(shape, generator=generator, device=device, dtype=dtype)
+
+        # get latents
+        try:
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_img2img.py:149]: need to make sure either it works with ddpm or add `add_noise` method to unclip scheduler

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_inpainting.py [hunk 14]
```diff
@@ -0,0 +1,481 @@
...
+class Kandinsky2_2_DecoderInpaintPipeline(DiffusionPipeline):
+    """
+    Pipeline for text-to-image generation using Kandinsky
+    This model inherits from [`DiffusionPipeline`]. Check the superclass documentation for the generic methods the
+    library implements for all the pipelines (such as downloading or saving, running on a particular device, etc.)
+    Args:
+        text_encoder ([`MultilingualCLIP`]):
+            Frozen text-encoder.
+        tokenizer ([`XLMRobertaTokenizer`]):
+            Tokenizer of class
```
Comment (yiyixuxu) [inline src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder_inpainting.py:194]:
```suggestion
```

**Review by yiyixuxu — COMMENTED** [review]
great! I left some feedbacks, most nit, we can wait to make changes until @patrickvonplaten gives his review 

I think the main to-dos left are:
- add docstring examples: I've tested our doc example for text2img, img2img and inpaint, all works great. only ones missing are controlnets 
- add `#copy from` statement everywhere (YiYi can help with this)
- add tests and doc (YiYi can help with this)
- make `PriorEmb2Emb` work as intended - either add add_noise function to unclip or replace unclip scheduler with ddpm (YiYi can help with this)

File: src/diffusers/models/embeddings.py [hunk 15]
```diff
@@ -375,6 +375,27 @@ def forward(self, text_embeds: torch.FloatTensor, image_embeds: torch.FloatTenso
...
+        self,
+        image_embed_dim: int = 768,
+        cross_attention_dim: int = 768,
+        num_image_text_embeds: int = 32,
+    ):
+        super().__init__()
+
+        self.num_image_text_embeds = num_image_text_embeds
+        self.image_embeds = nn.Linear(image_embed_dim, self.num_image_text_embeds * cross_attention_dim)
+        self.norm = nn.LayerNorm(cross_attention_dim)
```
Comment (patrickvonplaten) [inline src/diffusers/models/embeddings.py:389]:
```suggestion
        self.norm = nn.LayerNorm(cross_attention_dim)
        
```

File: src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder.py [hunk 16]
```diff
@@ -0,0 +1,320 @@
...
+            Conditional U-Net architecture to denoise the image embedding.
+        vae ([`VQModel`]):
+            MoVQ Decoder to generate the image from the latents.
+    """
+
+    def __init__(
+        self,
+        unet: UNet2DConditionModel,
+        scheduler: DDPMScheduler,
+        vae: VQModel,
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/pipelines/kandinsky2_2/pipeline_kandinsky2_2_decoder.py:87]: @cene555 is the model a movq here or just a vae?

File: src/diffusers/models/unet_2d_condition.py [hunk 17]
```diff
@@ -26,7 +26,10 @@
 from .embeddings import (
```
Comment (patrickvonplaten) [inline src/diffusers/models/unet_2d_condition.py:26]: Changes here look all good to me! would just be nice / important to add some tests and docstrings

File: src/diffusers/models/embeddings.py [hunk 18]
```diff
@@ -428,6 +449,46 @@ def forward(self, text_embeds: torch.FloatTensor, image_embeds: torch.FloatTenso
...
+                    nn.Conv2d(32, 32, 3, padding=1),
+                    nn.SiLU(),
+                    nn.Conv2d(32, 96, 3, padding=1, stride=2),
+                    nn.SiLU(),
+                    nn.Conv2d(96, 96, 3, padding=1),
+                    nn.SiLU(),
+                    nn.Conv2d(96, 256, 3, padding=1, stride=2),
+                    nn.SiLU(),
+                    nn.Conv2d(256, 4, 3, padding=1)
+                )
```
Comment (patrickvonplaten) [inline src/diffusers/models/embeddings.py:485]:
```suggestion
        self.input_hint_block = nn.Sequential(
            nn.Conv2d(3, 16, 3, padding=1),
            nn.SiLU(),
            nn.Conv2d(16, 16, 3, padding=1),
            nn.SiLU(),
            nn.Conv2d(16, 32, 3, padding=1, stride=2),
            nn.SiLU(),
            nn.Conv2d(32, 32, 3, padding=1),
            nn.SiLU(),
            nn.Conv2d(32, 96, 3, padding=1, stride=2),
            nn.SiLU(),
            nn.Conv2d(96, 96, 3, padding=1),
            nn.SiLU(),
            nn.Conv2d(96, 256, 3, padding=1, stride=2),
            nn.SiLU(),
            nn.Conv2d(256, 4, 3, padding=1)
        )
```

File: src/diffusers/models/embeddings.py [hunk 19]
```diff
@@ -375,6 +375,27 @@ def forward(self, text_embeds: torch.FloatTensor, image_embeds: torch.FloatTenso
```
Comment (patrickvonplaten) [inline src/diffusers/models/embeddings.py:375]: Changes look good to me!

File: src/diffusers/__init__.py [hunk 20]
```diff
@@ -136,6 +136,13 @@
         KandinskyInpaintPipeline,
         KandinskyPipeline,
         KandinskyPriorPipeline,
+        Kandinsky2_2_DecoderControlnetImg2ImgPipeline,
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply src/diffusers/__init__.py:139]: Yes I think it would be nicer to remove `_Decoder` here

Comment (patrickvonplaten) [inline src/diffusers/__init__.py:139]:
```suggestion
        Kandinsky2_2ControlnetImg2ImgPipeline,
```

File: src/diffusers/__init__.py [hunk 21]
```diff
@@ -136,6 +136,13 @@
...
+        Kandinsky2_2_DecoderControlnetPipeline,
+        Kandinsky2_2_DecoderImg2ImgPipeline,
+        Kandinsky2_2_DecoderPipeline,
+        Kandinsky2_2PriorEmb2EmbPipeline,
+        Kandinsky2_2PriorPipeline,
+        Kandinsky2_2_DecoderInpaintPipeline,
```
Comment (patrickvonplaten) [inline src/diffusers/__init__.py:145]:
```suggestion
        KandinskyV22ControlnetPipeline,
        KandinskyV22Img2ImgPipeline,
        KandinskyV22DecoderPipeline,
        KandinskyV22PriorEmb2EmbPipeline,
        KandinskyV22PriorPipeline,
        KandinskyV22InpaintPipeline,
```
